src/sig_stfl/xmss/sig_stfl_xmss_secret_key_functions.c

```diff
@@ -0,0 +1,42 @@
+// SPDX-License-Identifier: MIT
+
+#include <oqs/oqs.h>
+#include <string.h>
+#include "sig_stfl_xmss.h"
+
+/* Serialize XMSS secret key data into a byte string */
+OQS_STATUS OQS_SECRET_KEY_XMSS_serialize_key(const OQS_SIG_STFL_SECRET_KEY *sk, size_t *sk_len, uint8_t **sk_buf_ptr) {
+	if (sk == NULL || sk_len == NULL || sk_buf_ptr == NULL) {
+		return OQS_ERROR;
+	}
+
+	uint8_t *sk_buf = malloc(sk->length_secret_key * sizeof(uint8_t));
+	if (sk_buf == NULL) {
+		return OQS_ERROR;
+	}
+
+	// Simply copy byte string of secret_key_data
+	memcpy(sk_buf, sk->secret_key_data, sk->length_secret_key);
+
+	*sk_buf_ptr = sk_buf;
+	*sk_len = sk->length_secret_key;
+
+	return OQS_SUCCESS;
+}
+
+/* Deserialize XMSS byte string into an XMSS secret key data */
+OQS_STATUS OQS_SECRET_KEY_XMSS_deserialize_key(OQS_SIG_STFL_SECRET_KEY *sk, const size_t sk_len, const uint8_t *sk_buf) {
+	if (sk == NULL || sk_buf == NULL || (sk_len == 0) || (sk_len != sk->length_secret_key)) {
+		return OQS_ERROR;
+	}
+
+	if (sk->secret_key_data) {
+		// Key data already present
+		// We dont want to trample over data
+		return OQS_ERROR;
+	}
+
+	memcpy(sk->secret_key_data, sk_buf, sk_len);
+
+	return OQS_SUCCESS;
```

Comment:
So I have this piece of code that we discuss in the meeting: 
So we check if the address of array `secret_key_data` is either NULL or not. If it's not NULL, we assume it's already initialized, thus we return error. 
To reach line 39, `sk->secret_key_data` must be NULL. Which make the `memcpy` is SEGFAULT. 

I'm thinking of creating `static once = 1` as static variable to make sure the initialization happen only once. 
But this lead to:

- We will have 2 ways to allocate memory for `sk->secret_key_data`, either in `deserialize`, and in key generation.
- If I use `static` variable, I don't know in the case we multithread/fork process, would it become complicate to make sure the `sk->secret_key_data` happen only once with `static` variable.

Reply:
After a while I change the logic of the code to:

```c
/* Deserialize XMSS byte string into an XMSS secret key data */
OQS_STATUS OQS_SECRET_KEY_XMSS_deserialize_key(OQS_SIG_STFL_SECRET_KEY *sk, const size_t sk_len, const uint8_t *sk_buf) {
	if (sk == NULL || sk_buf == NULL || (sk_len != sk->length_secret_key)) {
		return OQS_ERROR;
	}

	if (sk->secret_key_data != NULL) {
		// Key data already present
		// We dont want to trample over data
		return OQS_ERROR;
	}

	sk->secret_key_data = malloc(sk_len);

	if (sk->secret_key_data == NULL) {
		return OQS_ERROR;
	}

	memcpy(sk->secret_key_data, sk_buf, sk_len);

	return OQS_SUCCESS;
}
```

What do you think @ashman-p @dstebila ?

Reply:
Commited in https://github.com/open-quantum-safe/liboqs/pull/1542/commits/7e4f2c93ddd90900320e1f74163efbe23938fc4a

Reply:
This looks fine to me.
It might be good to walk through the use-case where it is necessary to call "deserialize' on a working secret key object. Those use-cases would help determine the desired behaviors for each expected outcome.